keps/sig-cluster-lifecycle/0035-20190128-addons-via-operators.md

```diff
@@ -0,0 +1,181 @@
+KEP: Addons via Operators
+
+---
+kep-number: 35
+title: Addons via Operators
+authors:
+  - "@justinsb"
+owning-sig: sig-cluster-lifecycle
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-01-28
+last-updated: 2019-01-28
+status: provisional
+---
+
+# Addons via Operators
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Infrastructure Needed](#infrastructure-needed)
+
+
+## Summary
+
+We propose to use operators for managing cluster addons.  Each addon will have
```

Comment:
One thing I struggle with is: do we need a whole CRD for every addon.  Or is kustomize + default bundle of YAML sufficient?  I'm guessing this would depend on the addon.

Reply:
CRDs are supposed to be cheap, so I'd say we should have one for every addon - we do expect there will be per-addon-settings, and so a CRD-per-addon lets us actually use all this machinery we've invested so much in.

The hope is that we can make developing the controllers as easy as kustomize + default bundle of YAML though!

I'll add a comment to this effect

Reply:
Does this mean we'll have e.g. `/apis/addons.k8s.io/v1/kubeproxies` and `/apis/addons.k8s.io/v1/corednses` and so on?

Reply:
Yes (modulo naming and pluralization rules)

Reply:
Does it seem a bit weird to do that for singleton addons?

Reply:
Not really to my mind - can you clarify why it's weird?

I do think we should establish patterns for singletons (e.g. "name the instance default").  But I love the progress made on CRDs and think we should make use of them!

Reply:
we have followed a convention similar to what @justinsb described.  the operator basically only respects a single instance of a named cluster scoped CR for certain types of operators (not all).

Reply:
A naming pattern like `default` works for me.

Reply:
Maybe I'm missing something here, but having a different CRD type for each oddon makes more difficult to create tools for generic operations, like listing all addons with a given status, or watch for the installation of addos? What approaches should be used in that cases?